types/run.sh

```diff
@@ -78,22 +78,54 @@ checkable-files() {
 }
 
 need-typechecking() {
-    # This command is useful to find files to annotate and add to
-    # $MORE_OIL_MANIFEST.
-    # It shows all the files that are not included in
-    # $MORE_OIL_MANIFEST or $OSH_PARSE_MANIFEST, and thus are not yet
-    # typechecked by typecheck-more-oil here or
-    # `types/osh_parse.sh travis`.
-    comm -2 -3 \
-         <(checkable-files | sort | grep '.py$' | sed 's@^@./@') \
-         <(cat $MORE_OIL_MANIFEST $OSH_PARSE_MANIFEST | sort) \
-        | xargs wc -l | sort -n
+  # This command is useful to find files to annotate and add to
+  # $MORE_OIL_MANIFEST.
+  # It shows all the files that are not included in
+  # $MORE_OIL_MANIFEST or $OSH_PARSE_MANIFEST, and thus are not yet
+  # typechecked by typecheck-more-oil here or
+  # `types/osh_parse.sh travis`.
+  comm -2 -3 \
+    <(checkable-files | sort | grep '.py$' | sed 's@^@./@') \
+    <(cat $MORE_OIL_MANIFEST $OSH_PARSE_MANIFEST | sort) \
+    | xargs wc -l | sort -n
 }
 
 typecheck-files() {
   $0 typecheck --follow-imports=silent $MYPY_FLAGS "$@"
 }
 
+readonly -a COMMON_TYPE_MODULES=(_devbuild/gen/runtime_asdl.py _devbuild/gen/syntax_asdl.py)
+
+add-imports() {
+  # Temporary helper to add missing class imports to the 'if
+  # TYPE_CHECKING:' block of a single module, if the relevant
+  # classes are found in one of COMMON_TYPE_MODULES
+
+  # Also, this saves the typechecking output to the file named by
+  # $typecheck_out, to make it possible to avoid having to run two
+  # redundant (and slow) typechecking commands.  You can just cat that
+  # file after running this function.
+  local module=$1
+  export PYTHONPATH=.
+  readonly module_tmp=_tmp/add-imports-module.tmp
+  readonly typecheck_out=_tmp/add-imports-typecheck-output
+  set +o pipefail
+  # unbuffer is just to preserve colorization (it tricks the command
+  # into thinking it's writing to a pty instead of a pipe)
+  unbuffer types/run.sh typecheck-files "$module" | tee "$typecheck_out" | \
+    grep 'Name.*is not defined' | sed -r 's/.*'\''(\w+)'\''.*/\1/' | \
+    sort -u | python devtools/findclassdefs.py "${COMMON_TYPE_MODULES[@]}" | \
```

Comment:
Since these are just private scripts it's not worth messing with more, but I changed to use a shebang of `#!/usr/bin/env python2` and executing it directly (no `python`).  This will get Python 3 on some systems.

Reply:
Ooh! good point.  I've been using a python2 virtualenv for my oilshell work and didn't think of that.  Thanks.